lms/services/canvas_api/client.py

```diff
@@ -226,16 +227,30 @@ class _ListFilesSchema(RequestsResponseSchema):
         id = fields.Integer(required=True)
         updated_at = fields.String(required=True)
 
-    def public_url(self, file_id):
+    def public_url(self, course_id, file_id, is_instructor):
         """
         Get a new temporary public download URL for the file with the given ID.
 
+        :param course_id: the ID of the Canvas course
+        :rtype: str
+
         :param file_id: the ID of the Canvas file
         :rtype: str
+
+        :param is_instructor: Whether the current user is an instructor
+        :type is_instructor: bool
+
+        :raise CanvasFileNotFoundInCourse: If there's no file with ID file_id
+            in the course with ID course_id
         """
         # For documentation of this request see:
         # https://canvas.instructure.com/doc/api/files.html#method.files.public_url
 
+        if is_instructor:
+            file_ids = [str(file_["id"]) for file_ in self.list_files(course_id)]
+            if file_id not in file_ids:
+                raise CanvasFileNotFoundInCourse(file_id)
```

Comment:
We call the Canvas API to get a list of all the files in the course, with possible pagination and multiple requests, in order to determine whether the file is in the course.

We only want to do this for instructors because:

* We don't want to add these additional Canvas API calls to every Canvas Files assignment launch. Only to instructor launches
* The exception that we raise triggers and error page that tells the user to edit the assignment and re-select the file. Only instructors can do that
* We can't actually tell whether the file is in the course when a student launches an assignment anyway. The list files API will only list the files that the student has access to, which doesn't necessarily include all of the files in the course (instructors can control whether files are visible to students or not)

When a student launches an assignment if they don't have access to the file (which could be for a variety of reasons) we get a permissions error from the Canvas API and we show an error message to the student. That was done in a previous PR: https://github.com/hypothesis/lms/pull/2223

Reply:
`self.list_files(course_id)` is _not_ a generator